features/202-session-management/api.md

```diff
@@ -34,3 +48,19 @@ class AuthContainer {
     get currentSessionID(): string | null;
 }
 ```
+
+## SDK User Agent
+
+In general, for native SDKs, the user agent would have following format:
+```
+[App ID]/[App Version]([Device]; [OS]; [Device ID]) [SDK Library Name]/[SDK version]
+```
+
+Examples:
+- `io.skygear.test.ios/1.0.1 (iPhone11,8; iOS 12.0; 46FF79AA-4E2E-475F-B58F-E7E37BAB771C) SKYKit/2.0.1`
+- `io.skygear.test.android/1.3.0 (Samsung GT-S5830L; Android 9.0; 32DDE92083E1AA54) io.skygear.skygear/2.2.0`
```

Comment:
Is it really needed? Only browser have a "raw" agent string right? For other SDKs maybe they can just pass structured information to the server to fill in each data without a "raw" one?

Reply:
We should follow common convention of user agent string format (https://tools.ietf.org/html/rfc7231#section-5.5.3). I don't oppose sending an extra header containing structured data, but it would be redundant.

Reply:
I think the use of RFC7231's user agent here is strange, it defines the user agent in HTTP protocol, but for Skygear HTTPS is just the underlying protocol we used for APIs (So technically one day we could have a different RPC protocol), and than we use HTTPS's user agent as an attributes in our session data...

Reply:
Surely if we use other protocols (grpc?), we could send this information as a structured data. Since we just support HTTP at the moment, I think it's better to conform to the common convention of HTTP for now.

Reply:
However for now if you "overload" HTTP's user agent string of the network call between iOS / Android Client to the server, with a custom-made user agent string we made for user agent, it might leads to, say what if, for debug or statistic or whatever purpose, some developers think the raw user agent string our iOS's SDK HTTP library advertise itself as is useful? It is not available anymore as we already "overload" it with a user agent string of our own.

I don't think we are "conform to the common convention of HTTP for now", it is more like "a higher layer protocol overload a lower layer protocol's attributes for a special purpose"

Reply:
For your suggested case (`raw user agent string our iOS's SDK HTTP library advertise itself as is useful`), developer should not do this, since the HTTP library SDK used is not a public API and is subject to change.

If we used something like JSON for user agent, it would be indeed 'overloading a lower layer protocol's attributes'. The server is already assuming standard HTTP semantics, so I don't think it is our intention to do this?

Reply:
> For your suggested case (`raw user agent string our iOS's SDK HTTP library advertise itself as is useful`), developer should not do this, since the HTTP library SDK used is not a public API and is subject to change.

ya so I was saying maybe debugging or statistical purpose, or even the developer is writing its own clients (say Flutter client?) to connect with Skygear API



> If we used something like JSON for user agent, it would be indeed 'overloading a lower layer protocol's attributes'.

I have never said you should use a JSON for the HTTP user agent field. What about a custom header? 

BTW, I found that it is very confusing to call it user agent here, there are two user agent meaning: User Agent as in HTTP, or User Agent as in Sessions, maybe just make things clearer we should not call it User Agent in our sessions? Let's just use another name to avoid this confusion in future discussion / when people look at user agent in Skygear and thought it is 100% identical to the user agent in HTTP.

What about something like this:
```
interface Device {
    name: string; // browser name or app ID
    version: string; // browser version or app version
    os: string;
    osVersion: string;
    deviceType: string;
    http_user_agent: string;
}
```

Reply:
I don't see why developer would use user agent for debugging purpose? For statistical purpose, the proposed user agent string can be used. If developers would want to write SDK for other platform, they should confirm to the suggested user agent format so the server could parse meaningful information.

I don't agree there are two meanings of user agent: the user agent is the browser for web platform, and the app for native platform. User agent just means the application which the user used to interact with the server. The proposed user agent string formation reflect this clearly and conform to HTTP semantics.

Naming the user agent object as 'Device' would be misleading: it is mostly impossible to identify a device on modern platform.

Reply:
> I don't see why developer would use user agent for debugging purpose?

I don't know, maybe because they want to know which version of SDK is being used and sadly the only way they can identify it is different SDK's give different user agent string?

Again, what are the benefits of overloading HTTP user agent with our client's provided user agent information? 

> User agent just means the application which the user used to interact with the server.

Let's just use another name other than user agent, because it is confusing when most people use the term user agent, a developer would immediately think of HTTP user agent, but it is not in our case, we have mixed it up here already.

> Naming the user agent object as 'Device' would be misleading: it is mostly impossible to identify a device on modern platform.

This is just device information, calling it `Device` doesn't mean we must be able to uniquely identify a device. When you call it `User Agent` it also doesn't meant we can uniquely identify a user agent.

Reply:
> I don't know, maybe because they want to know which version of SDK is being used and sadly the only way they can identify it is different SDK's give different user agent string?

The proposed user agent includes the SDK name and version.

> This overloading have so much implication, as mentioned above, what if we have additional information that want to pass in the session's user agent/device attributes? In mobile SDK you can put it in, what about browser?

The user agent would contains strictly the information about the user agent. For browser, considering this: if the user spoof user agent using browser extension, do we still sniff the correct information in JS and send it to server? I would suggest using only the information available in user agent header for browser.

> Let's just use another name other than user agent, because it is confusing when most people use the term user agent, a developer would immediately think of HTTP user agent, but it is not in our case, we have mixed it up here already.

As mentioned, our case is indeed the HTTP user agent.

> This is just device information, calling it Device doesn't mean we must be able to uniquely identify a device. When you call it User Agent it also doesn't meant we can uniquely identify a user agent.

Fair enough. Still some fields would need to be renamed. (e.g. name -> user agent name, version -> user agent version)

Reply:
> The user agent would contains strictly the information about the user agent. For browser, considering this: if the user spoof user agent using browser extension, do we still sniff the correct information in JS and send it to server? I would suggest using only the information available in user agent header for browser.

it still doesn't cater for if we need to pass data other than those available in user agent.. say on the other thread we might have misled and need a custom name from each device for session data. Why we want to stop ourselves and make it really troubles to do it in future?

> As mentioned, our case is indeed the HTTP user agent.

As mentioned it is not -_- that is two different concept. I don't think we want to limit Skygear now to ONLY support HTTP as our transport protocol, I don't see any benefits we tie the two concepts together.



> Fair enough. Still some fields would need to be renamed. (e.g. name -> user agent name, version -> user agent version)

I don't think so? We can still just call it name and version.

Reply:
> The user agent would contains strictly the information about the user agent. For browser, considering this: if the user spoof user agent using browser extension, do we still sniff the correct information in JS and send it to server? I would suggest using only the information available in user agent header for browser.

BTW, that's why I said http user agent should only be ONE sources of information for Skygear's UserAgent/Device information.

Reply:
Summary of my opinions on how to do it:

1. I think we should use HTTP user-agent header as ONE of the sources of information. If we have additional information about the session's end device's information, we should use another way to send it to the server (e.g. a custom header)
   1. So for Mobile SDK, since many information won't be available in HTTP user-agent header, additional information should go to the custom header
1. In the API / data structure, we should store the HTTP user-agent as raw data, but also save the additional information or information extracted for the user-agent string.
1. Try to avoid the confusion of terms user-agent between HTTP user-agent (the HTTP client library) and Skygear's user-agent (the devices that's using the session)

Reply:
> I think we should use HTTP user-agent header as ONE of the sources of information...

Agree. For web, we use HTTP user-agent header. For native application, we use structured data the SDK collect from the device. If we choose to set the HTTP user-agent  in native application, then the value must conform to the HTTP RFC.

> In the API / data structure, we should store the HTTP user-agent as raw data, but also save the additional information or information extracted for the user-agent string.

Agree

Reply:
However, if we need to keep lastAccessedUserAgent up-to-date, the value should be transmitted via HTTP header. Add custom header(s) to represent structured data seems weird.

In case the client is not sending conforming HTTP user-agent (e.g. non-official flutter client), many information of `UserAgent` would be missing. So all fields should be made optional.

Reply:
Architecture-wise, we can do the following to safe-guard future implementation:

1. User Agent is one of the sources, especially for web SDK most information should be from it.
2. We have a custom header for additional information, which will override with the information we extract from http user agent header. 

So different client SDK implementation can do it with the "best way" for each client.



> However, if we need to keep lastAccessedUserAgent up-to-date, the value should be transmitted via HTTP header. Add custom header(s) to represent structured data seems weird.

Why? What is the werid part? @louischan-oursky  



> In case the client is not sending conforming HTTP user-agent (e.g. non-official flutter client), many information of `UserAgent` would be missing. So all fields should be made optional.

It is optional anyway, we don't even have anyway to "requires" client connecting to the server have a User Agent or any header... (unless we want it to be that strict to reject requests without a proper header?)

Reply:
> Why? What is the werid part? @louischan-oursky

Use a proprietary header while there is `user-agent`.

> We have a custom header for additional information, which will override with the information we extract from http user agent header.

Terry has proposed a format of `user-agent` of native application SDK. I guess we can have two identical headers `user-agent` and `x-skygear-user-agent` where the former can be overridden by the developer.

Reply:
> Use a proprietary header while there is `user-agent`.

@louischan-oursky I think my concern and different with you and @kiootic, as explained with the reasons above, is I don't think we should use the HTTP `user-agent` for our clients SDK to inject the information of sessions' device information. Instead, we should just treat the HTTP `user-agent` as one source of information, and inject additional information to a custom header (preferably structurally)

Let me put it into another perspective, imagine in a case where the client apps need to survive in an environment where a WAF filter all abnormal HTTP traffic, where the detection of abnormal includes some requirements on the HTTP User-Agent header. Isn't it weird for some information that should not appear in normal HTTP User-Agent string included only becoz we have magically injected some Skygear specific session device information in it?

> Terry has proposed a format of `user-agent` of native application SDK. I guess we can have two identical headers `user-agent` and `x-skygear-user-agent` where the former can be overridden by the developer.

So my suggestions is, instead we just put our session information into a custom header called something like `x-skygear-session-information` for the reason above.

The difference here should be minor, and I'm not 100% certain we will run into troubles with overriding HTTP user-agent in future.

But if we know there might be problem of `over-ridding` user-agent, why risk it? What is the downside of the alternative approach?

Reply:
I am fine with the approach proposed by Ben.

Reply:
Summarizing my opinions:
- In response to WAF: personally speaking, I'd say unknown header is much more abnormal than unknown user agent header value.
- According to RFC 7231, a user agent header should be included to indicated the user agent. Most HTTP library has already set user agent header to identify themselves. Therefore, regardless whether we store session info in it, client SDK should fill in user agent header correctly.
- In server-side rendered web app, a custom header cannot be sent with every request. A better approach would be a JS accessible cookie with a specified name, containing a JSON formatted extra device info. This feature should require user consent. (i.e. opt-in by developer by setting client side option).

Reply:
>     * In response to WAF: personally speaking, I'd say unknown header is much more abnormal than unknown user agent header value.

Not really in the age of XMLHTTPRequest all over the places...



>     * According to RFC 7231, a user agent header should be included to indicated the user agent. Most HTTP library has already set user agent header to identify themselves. Therefore, regardless whether we store session info in it, client SDK should fill in user agent header correctly.

Exactly -- HTTP library set user agent header to **identify themselves** as a HTTP client, not as a Skygear's Client...

>     * In server-side rendered web app, a custom header cannot be sent with every request. A better approach would be a JS accessible cookie with a specified name, containing a JSON formatted extra device info. This feature should require user consent. (i.e. opt-in by developer by setting client side option).

hmmm, but then for token-based sessions, we will have a header? So we will support both header and tokens even for client information? I think it is a bit over-complicated and maybe we could just drop sending extra Client's information for server-side rendered web app, feel like a okay trade-off for complexity (because for server-side rendered app, the client must be a Web Browser which provide enough information by its HTTP user-agent header already)

Reply:
> Exactly -- HTTP library set user agent header to identify themselves as a HTTP client, not as a Skygear's Client...

It seems you're think Skygear is something like a protocol implemented on top of HTTP protocol? If at the moment, we're using HTTP as the transport for API calls, the app (and the SDK) would be the user agent, and should be identified as such. For example, when gRPC is used with HTTP2 transport, it still recommend client libraries to fill in correct user agent header.

> hmmm, but then for token-based sessions, we will have a header? So we will support both header and tokens even for client information?

No, the extra info is put in a cookie regardless session token transport: the nature of this token is unrelated to authentication. It is expected that client SDK would have a cookie jar ready, in case server is configured to use cookie as session token transport, so using cookie for extra info would not be more complex.

Reply:
I think requires Client SDK must have a cookie jar ready even for token based session (which I believe is the default) is an additional technical complexity which doesn't worth for the feature of supplying session devices information, please don't do it.

Reply:
TBH, if I am forced to choose between accept overriding all sessions information in HTTP user-agent, or add a requirement of using cookies, I would definitely goes for reusing HTTP user-agent for the session information purpose XD

Reply:
> It seems you're think Skygear is something like a protocol implemented on top of HTTP protocol? If at the moment, we're using HTTP as the transport for API calls, the app (and the SDK) would be the user agent, and should be identified as such. For example, when gRPC is used with HTTP2 transport, it still recommend client libraries to fill in correct user agent header.

Yes, as mentioned above, we're just playing safe here -- since I don't see any real downside of not overriding user-agent and use a custom header, why not...

But anyway if you really really insist and we don't have further reasons, I'm okay to just accept it and use http user-agent, chance is 99% we won't run into any trouble anyway and it's not worth the time for such a long back-and-forth. I also hate we suddenly come up with requirements that force client SDK to have a cookie jar which is even more unacceptable.